pkg/util/atomicstats/reporter.go

```diff
@@ -0,0 +1,461 @@
+func NewReporter(v interface{}) Reporter {
+	ptrType := reflect.TypeOf(v)
+
+	if ptrType.Kind() != reflect.Ptr || ptrType.Elem().Kind() != reflect.Struct {
+		// errors here are programming (type) errors that cannot be caught at compile time,
+		// and thus are fatal
+		panic("NewReporter expects a pointer to a struct")
+	}
+
+	structType := ptrType.Elem()
+
+	reporter := Reporter{
+		structType: structType,
+		fields:     map[string]field{},
+	}
+
+	for idx := 0; idx < structType.NumField(); idx++ {
+		fieldType := structType.Field(idx)
+		if _, ok := fieldType.Tag.Lookup(statsTag); ok {
+			fld, err := getFieldFor(fieldType.Type)
+			if err != nil {
+				panic(err.Error())
+			}
+
+			fld.idx = idx
+			reporter.fields[toSnakeCase(fieldType.Name)] = fld
+		}
+	}
+
+	return reporter
+}
+
+// Report returns a `map` representation of the stats in the given value.
+// All map keys are converted to snake_case.
+//
+// The value _must_ have the same type as the value passed to NewReporter.
+func (r *Reporter) Report(v interface{}) map[string]interface{} {
+	value := reflect.ValueOf(v)
+	r.checkType(value)
+
+	// dereference the pointer
+	value = value.Elem()
+
+	stats := make(map[string]interface{}, len(r.fields))
+	for name, fld := range r.fields {
+		stats[name] = fld.report(value.Field(fld.idx))
+	}
+	return stats
+}
+
+// Reset resets all stats fields to their zero value and returns the same kind
+// of value as Report.
+//
+// Atomic values are read with `Swap` when that method is defined, so no values
+// are lost. Note that concurrent access to the stats struct can update values
+// while this method is running, even if the fields are atomic.  For example,
+// if the stats struct contains `goodHits`, `badHits`, and `allHits` counters,
+// it is not a valid assumption that `goodHits + badHits == allHits`, as some
+// hit may have been counted in `goodHits` before Reset read the field, but in
+// `allHits` after the Reset.
+//
+// Non-atomic values can only be reset if they are in an exported field (initial
+// uppercase letter), as go does not allow un-exported values to be written using
+// reflect.  Resetting a struct with unexported stats fields will panic.
+func (r *Reporter) Reset(v interface{}) map[string]interface{} {
+	value := reflect.ValueOf(v)
+	r.checkType(value)
+
+	// dereference the pointer
+	value = value.Elem()
+
+	stats := make(map[string]interface{}, len(r.fields))
+	for name, fld := range r.fields {
+		stats[name] = fld.reset(value.Field(fld.idx))
+	}
+	return stats
+}
+
+// checkType checks that the given value is a pointer to r.structType
+func (r *Reporter) checkType(value reflect.Value) {
+	if value.Type() != reflect.PtrTo(r.structType) {
+		panic(fmt.Sprintf("Reporter#Report expects %s; got %s", reflect.PtrTo(r.structType), value.Type()))
+	}
+}
+
+func getFieldFor(fieldType reflect.Type) (field, error) {
+	switch fieldType.Kind() {
+	case reflect.Int:
+		return field{
```

Comment:
Do these need to be allocated every time? You could potentially put these in a `map[reflect.Kind]field` on init and then return them here?